Mechanically port bulk of x86 builtins to TableGen

chandlerc opened the pull request:
The goal is to make incremental (if small) progress towards fully TableGen'ed builtins, and to unblock #120534 by gaining access to more powerful TableGen-based representations.

The bulk `.td` file addition was generated with the help of a very rough Python script. That script made no attempt to be robust or reusable, it specifically handled only the cases in the X86 `.def` file.

Four entries from the `.def` file were not handled automatically as they used `BUILTIN` rather than `TARGET_BUILTIN`. These were ported by hand to an empty-feature `TargetBuiltin` entry, which seems like a better match.

For all the automatically ported entries, the results were compared by sorting and diffing the `.def` file and the generated `.inc` file. The only differences were:

- Different horizontal whitespace

- Additional entries that had already been ported to the `.td` file.

- More systematically using `Oi` instead of `LLi` for the type `long long int` in the fully general `__builtin_ia32_...` builtins for OpenCL support. The `.def` file was only partially moved to this it seems, and the systematic migration has updated a few missed builtins.

chandlerc committed:
Mechanically port bulk of x86 builtins to TableGen

The goal is to make incremental (if small) progress towards fully
TableGen'ed builtins, and to unblock #120534 by gaining access to more
powerful TableGen-based representations.

The bulk `.td` file addition was generated with the help of a very rough
Python script. That script made no attempt to be robust or reusable, it
specifically handled only the cases in the X86 `.def` file.

Four entries from the `.def` file were not handled automatically as they
used `BUILTIN` rather than `TARGET_BUILTIN`. These were ported by hand
to an empty-feature `TargetBuiltin` entry, which seems like a better
match.

For all the automatically ported entries, the results were compared by
sorting and diffing the `.def` file and the generated `.inc` file. The
only differences were:

- Different horizontal whitespace

- Additional entries that had already been ported to the `.td` file.

- Systematically using `Oi` instead of `LLi` for the type `long long
  int`. The `.def` file uses a mixture of `Oi` and `LLi`. I chose the
  shorter encoding.

This gives me high confidence in the correctness of the change.

phoebewang commented:
> * Systematically using `Oi` instead of `LLi` for the type `long long int`. The `.def` file uses a mixture of `Oi` and `LLi`. I chose the shorter encoding.

The mixture use of `Oi` and `LLi` is a mess, but `Oi` has different meaning for OpenCL targets. I think we should not change `LLi` to `Oi`. I think a lot `Oi` can be replaced with `LLi`, but I cannot tell which are required with a quick look.

chandlerc commented:
> > * Systematically using `Oi` instead of `LLi` for the type `long long int`. The `.def` file uses a mixture of `Oi` and `LLi`. I chose the shorter encoding.
> 
> The mixture use of `Oi` and `LLi` is a mess, but `Oi` has different meaning for OpenCL targets. I think we should not change `LLi` to `Oi`. I think a lot `Oi` can be replaced with `LLi`, but I cannot tell which are required with a quick look.

I do understand that, but I'm actually a bit more confident that these changes are correct... Or at least not a regression.

Specifically, the x86 intrinsic builtins use `Oi` *very* consistently for "quad word" (64-bit) vector operations from SSE through AVX2. For example `__builtin_ia32_psllqi256` uses `V4Oi`. This seems quite intentional, so I was very hesitant to reverse it.

The places where `LLi` has crept in are:
- 1 or 2 very isolated cases I'll list explicitly below
- Some AVX-512 and SHA512 intrinsics. This seems like a mistake as they're also using it for "double word" (64-bit) vector elements, the exact same element size that uses `Oi` elsewhere. And some actually *do* use `Oi` with the same feature (`avx10.2-256`), so I couldn't see any pattern that seemed to indicate a critical distinction...

I think the AVX-512 stuff was just added without realizing the historical use of `Oi` here?

The only examples outside of AVX-512 and SHA512 I could find:
- `__builtin_ia32_rdpru`
- `__emul` and `__emulu` (MS intrinsics)
- `__readfsqword` and `__readgsqword` (also MS intrinsics)

I can understand that these don't make lots of sense in OpenCL, but they also seem very unlikely to show up or do the wrong thing here, when getting the same type that OpenCL uses for 64-bit vector elements on x86?

The reason I'd like to consolidate here is that preserving this distinction would add complexity to the tablegen code, and at least *seems* like it may be propagating more of a mistake than a careful choice in AVX-512 (especially given the lack of test coverage that errors with the change as is).

chandlerc commented:
A long way from an expert on OpenCL, but it seems to not even have the concept of `long long`, and `long` is defined as a 64-bit type (and just optional for embedded stuff)?

https://registry.khronos.org/OpenCL/sdk/3.0/docs/man/html/scalarDataTypes.html

phoebewang commented:
> A long way from an expert on OpenCL, but it seems to not even have the concept of `long long`, and `long` is defined as a 64-bit type (and just optional for embedded stuff)?
> 
> https://registry.khronos.org/OpenCL/sdk/3.0/docs/man/html/scalarDataTypes.html

Thanks for the confirmation! Does OpenCL supports Windows? IIRC, long is 32-bit on Windows.

topperc commented:
When I was still involved in X86 my recollection was we primarily used LLi. It looks like there was a large replacement of LLi with Oi here fa8cd7691ac28d07f6a127ed26f0dbe49699bd59.

chandlerc commented:
> > A long way from an expert on OpenCL, but it seems to not even have the concept of `long long`, and `long` is defined as a 64-bit type (and just optional for embedded stuff)?
> > 
> > https://registry.khronos.org/OpenCL/sdk/3.0/docs/man/html/scalarDataTypes.html
> 
> Thanks for the confirmation! Does OpenCL supports Windows? IIRC, long is 32-bit on Windows.

Yeah, it's definitely different on windows in C/C++.  My impression is that opencl long is definitively 64 bit, and thats kind of why Oi exists -- to map that to `long long` outside of opencl where it would be ambiguous otherwise.

chandlerc commented:
> When I was still involved in X86 my recollection was we primarily used LLi. It looks like there was a large replacement of LLi with Oi here fa8cd7691ac28d07f6a127ed26f0dbe49699bd59.

Yeah, this patch makes me think the change to `Oi` here is ultimately correct, and focuses types on the correct 64-bit integer type in the different languages.

chandlerc commented:
Ping, rebased to top-of-tree.

@phoebewang -- I think you're the most relevant reviewer here. If the `O` vs. `LL` thing is really a blocker despite the added information, I'd like to know so I can explore options to switch back. All of the ones I've come up with add complexity, so I'm hoping the current version is OK, but if its a blocker happy to start discussing what would work here.

I think the latest question was around MSVC / Windows where `long` in C and C++ is 32-bit, but AFAICT with my research in the standard, it is *always* 64-bits in OpenCL.

topperc commented on clang/utils/TableGen/ClangBuiltinsEmitter.cpp:
```diff
@@ -108,9 +109,15 @@ class PrototypeParser {
     } else if (T.consume_back("&")) {
       ParseType(T);
       Type += "&";
+    } else if (T.consume_front("long long")) {
```
It feels a bit surprising that "long long" wouldn't map to "LL". I get that we need to support OpenCL, but maybe we should have a keyword for OpenCL? Are all targets going to want "O" for "long long"?

chandlerc replied:
Not sure -- the vast majority of x86 builtins use `O` for this.

It's a no-op on targets without OpenCL support, but it seems harmless there.

chandlerc replied:
This does maybe point at something that doesn't add much complexity -- I can condition using `O` on a flag that only X86 builtins use so it doesn't impact any other targets? That should be quite simple, and then other targets can opt in if/when they wish?

chandlerc replied:
Ok, PR updated with an explicit opt-in for OpenCL `long` type support.

Somehow, I hadn't considered how easily this would address an unrelated part: the occurance in intrinsic header builtins. That just fell out of this. Sorry for pushing back earlier, but all my ideas were much more complex than this ended up being.

With this tiny change to the `.td` file in the second commit here, the diff of things switching from `LLi` to `Oi` becomes very small and looks pretty compelling: https://gist.github.com/chandlerc/4395df8d838cd1a110ecc2170e67adc4

phoebewang replied:
How did you get this diff? I think it's useful for us to verify the correctness.

chandlerc replied:
Sure, I use the Fish shell and have a bunch of command line tools that help with this installed:

- https://github.com/BurntSushi/ripgrep as `rg`, nicer form of `grep`
- https://github.com/chmln/sd as `sd`, nicer form of `sed` for line-based regex search and replace

Combined, they let me write a `diff` command like:

```fish
diff -u (rg -I '^(TARGET|BUILTIN)' BuiltinsX86.def BuiltinsX86.inc | sd '^BUILTIN\((.*)\)' 'TARGET_BUILTIN($1, "")' | sd ' +' ' ' | sd ',([X"])' ', $1' | sort | psub)  (rg '^TARGET' dev/tools/clang/include/clang/Basic/BuiltinsX86.inc | sort | psub)
```

Here `BuiltinsX86.def` and `BuiltinsX86.inc` are copies of the `.def` and `.inc` from before this PR, and `dev/tools/.../BuiltinsX86.inc` is the `.inc` produced in my development build.

All the uses of `sd` on the original `.def` and `.inc` files are to process away weird whitespace artifacts, and the use of `BUILTIN` instead of `TARGET_BUILTIN` in four places.

phoebewang replied:
Thanks for the information!

chandlerc committed:
Make the OpenCL type opt-in, and only use it for non-header and non-lib
X86 builtins.

This minimizes the delta from the non-TableGen and avoids unintended
consequences on other targets.

phoebewang reviewed (approved):
LGTM.

chandlerc commented:
Thanks, merging! I've put the script here for posterity: https://gist.github.com/chandlerc/de807ea073beac351f87c660e1d4b7a0

chandlerc commented:
> LLVM Buildbot has detected a new failure on builder `clang-debian-cpp20` running on `clang-debian-cpp20` while building `clang` at step 6 "test-build-unified-tree-check-all".
> 
> Full details are available at: https://lab.llvm.org/buildbot/#/builders/108/builds/7722
> 
> <details>
> <summary>Here is the relevant piece of the build log for the reference</summary>
> 
> ```
> Step 6 (test-build-unified-tree-check-all) failure: test (failure)
> ******************** TEST 'LLVM :: ExecutionEngine/JITLink/RISCV/ELF_riscv64_got_plt_reloc.s' FAILED ********************
> Exit Code: 134
> 
> Command Output (stderr):
> --
> RUN: at line 1: rm -rf /vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/test/ExecutionEngine/JITLink/RISCV/Output/ELF_riscv64_got_plt_reloc.s.tmp && mkdir -p /vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/test/ExecutionEngine/JITLink/RISCV/Output/ELF_riscv64_got_plt_reloc.s.tmp
> + rm -rf /vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/test/ExecutionEngine/JITLink/RISCV/Output/ELF_riscv64_got_plt_reloc.s.tmp
> + mkdir -p /vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/test/ExecutionEngine/JITLink/RISCV/Output/ELF_riscv64_got_plt_reloc.s.tmp
> RUN: at line 2: /vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/bin/llvm-mc -triple=riscv64 -position-independent -filetype=obj      -o /vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/test/ExecutionEngine/JITLink/RISCV/Output/ELF_riscv64_got_plt_reloc.s.tmp/elf_riscv64_got_plt_reloc.o /vol/worker/clang-debian-cpp20/clang-debian-cpp20/llvm-project/llvm/test/ExecutionEngine/JITLink/RISCV/ELF_riscv64_got_plt_reloc.s
> + /vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/bin/llvm-mc -triple=riscv64 -position-independent -filetype=obj -o /vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/test/ExecutionEngine/JITLink/RISCV/Output/ELF_riscv64_got_plt_reloc.s.tmp/elf_riscv64_got_plt_reloc.o /vol/worker/clang-debian-cpp20/clang-debian-cpp20/llvm-project/llvm/test/ExecutionEngine/JITLink/RISCV/ELF_riscv64_got_plt_reloc.s
> RUN: at line 4: /vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/bin/llvm-jitlink -noexec      -slab-allocate 100Kb -slab-address 0xfff00000 -slab-page-size 4096      -abs external_func=0x1 -abs external_data=0x2      -check /vol/worker/clang-debian-cpp20/clang-debian-cpp20/llvm-project/llvm/test/ExecutionEngine/JITLink/RISCV/ELF_riscv64_got_plt_reloc.s /vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/test/ExecutionEngine/JITLink/RISCV/Output/ELF_riscv64_got_plt_reloc.s.tmp/elf_riscv64_got_plt_reloc.o
> + /vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/bin/llvm-jitlink -noexec -slab-allocate 100Kb -slab-address 0xfff00000 -slab-page-size 4096 -abs external_func=0x1 -abs external_data=0x2 -check /vol/worker/clang-debian-cpp20/clang-debian-cpp20/llvm-project/llvm/test/ExecutionEngine/JITLink/RISCV/ELF_riscv64_got_plt_reloc.s /vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/test/ExecutionEngine/JITLink/RISCV/Output/ELF_riscv64_got_plt_reloc.s.tmp/elf_riscv64_got_plt_reloc.o
> RUN: at line 10: /vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/bin/llvm-mc -triple=riscv64 -position-independent -filetype=obj      -mattr=+relax -o /vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/test/ExecutionEngine/JITLink/RISCV/Output/ELF_riscv64_got_plt_reloc.s.tmp/elf_riscv64_got_plt_reloc.o /vol/worker/clang-debian-cpp20/clang-debian-cpp20/llvm-project/llvm/test/ExecutionEngine/JITLink/RISCV/ELF_riscv64_got_plt_reloc.s
> + /vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/bin/llvm-mc -triple=riscv64 -position-independent -filetype=obj -mattr=+relax -o /vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/test/ExecutionEngine/JITLink/RISCV/Output/ELF_riscv64_got_plt_reloc.s.tmp/elf_riscv64_got_plt_reloc.o /vol/worker/clang-debian-cpp20/clang-debian-cpp20/llvm-project/llvm/test/ExecutionEngine/JITLink/RISCV/ELF_riscv64_got_plt_reloc.s
> RUN: at line 12: /vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/bin/llvm-jitlink -noexec      -slab-allocate 100Kb -slab-address 0xfff00000 -slab-page-size 4096      -abs external_func=0x1 -abs external_data=0x2      -check /vol/worker/clang-debian-cpp20/clang-debian-cpp20/llvm-project/llvm/test/ExecutionEngine/JITLink/RISCV/ELF_riscv64_got_plt_reloc.s /vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/test/ExecutionEngine/JITLink/RISCV/Output/ELF_riscv64_got_plt_reloc.s.tmp/elf_riscv64_got_plt_reloc.o
> + /vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/bin/llvm-jitlink -noexec -slab-allocate 100Kb -slab-address 0xfff00000 -slab-page-size 4096 -abs external_func=0x1 -abs external_data=0x2 -check /vol/worker/clang-debian-cpp20/clang-debian-cpp20/llvm-project/llvm/test/ExecutionEngine/JITLink/RISCV/ELF_riscv64_got_plt_reloc.s /vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/test/ExecutionEngine/JITLink/RISCV/Output/ELF_riscv64_got_plt_reloc.s.tmp/elf_riscv64_got_plt_reloc.o
> llvm-jitlink: /vol/worker/clang-debian-cpp20/clang-debian-cpp20/llvm-project/llvm/include/llvm/ExecutionEngine/Orc/SymbolStringPool.h:285: llvm::orc::SymbolStringPool::~SymbolStringPool(): Assertion `Pool.empty() && "Dangling references at pool destruction time"' failed.
> PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
> Stack dump:
> 0.	Program arguments: /vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/bin/llvm-jitlink -noexec -slab-allocate 100Kb -slab-address 0xfff00000 -slab-page-size 4096 -abs external_func=0x1 -abs external_data=0x2 -check /vol/worker/clang-debian-cpp20/clang-debian-cpp20/llvm-project/llvm/test/ExecutionEngine/JITLink/RISCV/ELF_riscv64_got_plt_reloc.s /vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/test/ExecutionEngine/JITLink/RISCV/Output/ELF_riscv64_got_plt_reloc.s.tmp/elf_riscv64_got_plt_reloc.o
>  #0 0x0000597f47dc4b98 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/bin/llvm-jitlink+0xeb5b98)
>  #1 0x0000597f47dc268d llvm::sys::RunSignalHandlers() (/vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/bin/llvm-jitlink+0xeb368d)
>  #2 0x0000597f47dc5138 SignalHandler(int) Signals.cpp:0:0
>  #3 0x00007a5a62d4d510 (/lib/x86_64-linux-gnu/libc.so.6+0x3c510)
>  #4 0x00007a5a62d9b0fc (/lib/x86_64-linux-gnu/libc.so.6+0x8a0fc)
>  #5 0x00007a5a62d4d472 raise (/lib/x86_64-linux-gnu/libc.so.6+0x3c472)
>  #6 0x00007a5a62d374b2 abort (/lib/x86_64-linux-gnu/libc.so.6+0x264b2)
>  #7 0x00007a5a62d373d5 (/lib/x86_64-linux-gnu/libc.so.6+0x263d5)
>  #8 0x00007a5a62d463a2 (/lib/x86_64-linux-gnu/libc.so.6+0x353a2)
>  #9 0x0000597f4765df0c (/vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/bin/llvm-jitlink+0x74ef0c)
> #10 0x0000597f47ca7b37 llvm::orc::ExecutorProcessControl::~ExecutorProcessControl() (/vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/bin/llvm-jitlink+0xd98b37)
> #11 0x0000597f47ca927f llvm::orc::SelfExecutorProcessControl::~SelfExecutorProcessControl() (/vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/bin/llvm-jitlink+0xd9a27f)
> #12 0x0000597f47bdae88 llvm::orc::ExecutionSession::~ExecutionSession() (/vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/bin/llvm-jitlink+0xccbe88)
> #13 0x0000597f47636edd llvm::Session::~Session() (/vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/bin/llvm-jitlink+0x727edd)
> #14 0x0000597f47641664 main (/vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/bin/llvm-jitlink+0x732664)
> #15 0x00007a5a62d386ca (/lib/x86_64-linux-gnu/libc.so.6+0x276ca)
> #16 0x00007a5a62d38785 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x27785)
> #17 0x0000597f4762f671 _start (/vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/bin/llvm-jitlink+0x720671)
> /vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/test/ExecutionEngine/JITLink/RISCV/Output/ELF_riscv64_got_plt_reloc.s.script: line 5: 3606366 Aborted                 /vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/bin/llvm-jitlink -noexec -slab-allocate 100Kb -slab-address 0xfff00000 -slab-page-size 4096 -abs external_func=0x1 -abs external_data=0x2 -check /vol/worker/clang-debian-cpp20/clang-debian-cpp20/llvm-project/llvm/test/ExecutionEngine/JITLink/RISCV/ELF_riscv64_got_plt_reloc.s /vol/worker/clang-debian-cpp20/clang-debian-cpp20/build/test/ExecutionEngine/JITLink/RISCV/Output/ELF_riscv64_got_plt_reloc.s.tmp/elf_riscv64_got_plt_reloc.o
> 
> --
> 
> ********************
> 
> 
> ```
> 
> </details>

This is an LLVM failure and so I can't see how it relates. Likely a flaky crash of a tool given the log.
